vowpalwabbit/core/include/vw/core/thread_pool.h

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+#pragma once
+
+#include "vw/core/memory.h"
+
+#include <atomic>
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+// thread pool based off of the simple pool implementation from ccia:
+// https://github.com/anthonywilliams/ccia_code_samples
+
+namespace VW
+{
+template <typename T>
+class thread_safe_queue
+{
+public:
+  thread_safe_queue() {}
+  void push(T value)
+  {
+    std::lock_guard<std::mutex> lk(_mut);
+    _queue.push(std::move(value));
+  }
+
+  bool try_pop(T& value)
+  {
+    std::lock_guard<std::mutex> lk(_mut);
+    if (_queue.empty()) return false;
+    value = std::move(_queue.front());
+    _queue.pop();
+    return true;
+  }
+
+  bool empty() const
+  {
+    std::lock_guard<std::mutex> lk(_mut);
+    return _queue.empty();
+  }
+
+private:
+  mutable std::mutex _mut;
+  std::queue<T> _queue;
+};
+
+class threads_joiner
+{
+public:
+  explicit threads_joiner(std::vector<std::thread>& threads) : _threads{threads} {}
+  ~threads_joiner()
+  {
+    for (std::thread& thread : _threads)
+    {
+      if (thread.joinable()) thread.join();
+    }
+  }
+
+private:
+  std::vector<std::thread>& _threads;
+};
+
+class thread_pool
+{
+private:
+  class function_wrapper
+  {
+    // function wrapper is a type erased move-only function storer needed in order to temporarily store the submitted
+    // task
+    // in a std::packaged_task which in turn is move-only
+  public:
+    template <typename F>
+    function_wrapper(F&& f) : _impl{VW::make_unique<impl_type<F>>(std::forward<F>(f))}
+    {
+    }
+
+    function_wrapper() = default;
+
+    function_wrapper(function_wrapper&& other) : _impl{std::move(other._impl)} {}
+
+    function_wrapper(const function_wrapper&) = delete;
+
+    function_wrapper(function_wrapper&) = delete;
+
+    function_wrapper& operator=(const function_wrapper&) = delete;
+
+    void operator()() { _impl->call(); }

```

Comment:
It looks like this means the thread pool cant be used for operations which return values?